lib/puppet/parser/resource.rb

```diff
@@ -199,6 +199,51 @@ def raw_tagged?(tag_array)
     super || ((scope_resource = scope.resource) && !scope_resource.equal?(self) && scope_resource.raw_tagged?(tag_array))
   end
 
+  # Fill resource params from a capability
+  #
+  # This backs 'consumes => Sql[one]'
+  def add_parameters_from_consume
+    return unless self[:consume]
+
+    map = {}
+    [ self[:consume] ].flatten.map do |ref|
+      # Resolve references
+      catalog.resource(ref.type, ref.title) or
+        raise "Resource #{ref} could not be found; it might not have been produced yet"
+    end.each do |cns|
+      # Establish mappings
+      blueprint = resource_type.consumes.find do |bp|
+        bp[:capability] == cns.type
+      end
+      # @todo lutter 2015-08-03: catch this earlier, can we do this during
+      # static analysis ?
+      raise "Resource #{self} tries to consume #{cns} but no 'consumes' mapping exists for #{self.resource_type} and #{cns.type}" unless blueprint
+
+      # setup scope that has, for each attr of cns, a binding to cns[attr]
+      cns_scope = scope.find_global_scope.newscope(:source => self, :resource => self)
+      cns.to_hash.each { |name, value| cns_scope[name.to_s] = value }
+
+      # evaluate mappings in that scope
+      resource_type.arguments.keys.each do |name|
+        if expr = blueprint[:mappings][name]
+          # Explicit mapping
+          value = expr.safeevaluate(cns_scope)
+        else
+          value = cns[name]
+        end
+        if value
+          # @todo lutter 2015-07-01: this should be caught by the checker
+          # much earlier. We consume several capres, at least two of which
+          # want to map to the same parameter (APPMGMT-760)
+          raise "Several consumed mappings try to set #{name}" if map[name]
+          map[name] = value
+        end
+      end
+    end
+
+    map.each { |name, value| self[name] = value unless self[name] }
```

Comment:
this will probably also have problems with boolean values, the unless part should probably explicitly test against non nil. `self[name] = value if self[name].nil?`